src/compiler/types.ts

```diff
@@ -3472,39 +3488,42 @@ namespace ts {
         ContainsTypeScript = 1 << 1,
         Jsx = 1 << 2,
         ContainsJsx = 1 << 3,
-        ES2017 = 1 << 4,
-        ContainsES2017 = 1 << 5,
-        ES2016 = 1 << 6,
-        ContainsES2016 = 1 << 7,
-        ES2015 = 1 << 8,
-        ContainsES2015 = 1 << 9,
-        Generator = 1 << 10,
-        ContainsGenerator = 1 << 11,
-        DestructuringAssignment = 1 << 12,
-        ContainsDestructuringAssignment = 1 << 13,
+        ESNext = 1 << 4,
+        ContainsESNext = 1 << 5,
+        ES2017 = 1 << 6,
+        ContainsES2017 = 1 << 7,
+        ES2016 = 1 << 8,
+        ContainsES2016 = 1 << 9,
+        ES2015 = 1 << 10,
+        ContainsES2015 = 1 << 11,
+        Generator = 1 << 12,
+        ContainsGenerator = 1 << 13,
+        DestructuringAssignment = 1 << 14,
+        ContainsDestructuringAssignment = 1 << 15,
 
         // Markers
         // - Flags used to indicate that a subtree contains a specific transformation.
-        ContainsDecorators = 1 << 14,
-        ContainsPropertyInitializer = 1 << 15,
-        ContainsLexicalThis = 1 << 16,
-        ContainsCapturedLexicalThis = 1 << 17,
-        ContainsLexicalThisInComputedPropertyName = 1 << 18,
-        ContainsDefaultValueAssignments = 1 << 19,
-        ContainsParameterPropertyAssignments = 1 << 20,
-        ContainsSpreadElementExpression = 1 << 21,
-        ContainsComputedPropertyName = 1 << 22,
-        ContainsBlockScopedBinding = 1 << 23,
-        ContainsBindingPattern = 1 << 24,
-        ContainsYield = 1 << 25,
-        ContainsHoistedDeclarationOrCompletion = 1 << 26,
+        ContainsDecorators = 1 << 16,
+        ContainsPropertyInitializer = 1 << 17,
+        ContainsLexicalThis = 1 << 18,
+        ContainsCapturedLexicalThis = 1 << 19,
+        ContainsLexicalThisInComputedPropertyName = 1 << 20,
+        ContainsDefaultValueAssignments = 1 << 21,
+        ContainsParameterPropertyAssignments = 1 << 22,
+        ContainsSpreadExpression = 1 << 23,
+        ContainsComputedPropertyName = 1 << 24,
+        ContainsBlockScopedBinding = 1 << 25,
+        ContainsBindingPattern = 1 << 26,
+        ContainsYield = 1 << 27,
+        ContainsHoistedDeclarationOrCompletion = 1 << 28,

```

Comment:
I think I can avoid the over-emit inside destructuring by checking whether the destructuring is happening for ESNext. If it's ESNext, then I can just do nothing to nodes that have ContainsSpreadElement if they are array rest rather than object rest.